Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending, AddProduct (var productItem, _)) =>
+            [
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending pending, RemoveProduct((var productId, var toRemove, var price) pricedProductItem)) =>
+                pending.ProductItems.TryGetValue($"{productId}{price}", out var quantity) && quantity >= toRemove
+                    ?
+                    [
+                        new ProductRemoved(pricedProductItem, Now)
+                    ]
+                    : throw new InvalidOperationException("Not enough Products!"),
+
+            (Pending, Confirm(var clientId)) =>
+            [
+                new Confirmed(clientId, Now)
+            ],
+
+            (Pending, Cancel) =>
+            [
+                new Cancelled(Now)
+            ],
+
+            (Closed, Confirm) => [],
+
+            (Closed, Cancel) => [],
+
+            _ => throw new InvalidOperationException(nameof(command))
+        };
+
+    private DateTimeOffset Now => timeProvider.GetLocalNow();
+}
+
+public class ShoppingCartModel
+{
+    public required Guid Id { get; init; }
+    public Guid? ClientId { get; set; }
+    public required ShoppingCartStatus Status { get; set; }
+    public List<PricedProductItem> ProductItems { get; init; } = [];
+    public required DateTimeOffset OpenedAt { get; set; }
+    public DateTimeOffset? ConfirmedAt { get; set; }
+    public DateTimeOffset? CanceledAt { get; set; }
+
+    public class PricedProductItem
+    {
+        public required Guid ProductId { get; init; }
+        public required int Quantity { get; set; }
+        public required decimal Price { get; init; }
+        public required DateTimeOffset AddedAt { get; set; }
+        public DateTimeOffset? UpdatedAt { get; set; }
+    }
+
+    public enum ShoppingCartStatus
+    {
+        Pending,
+        Confirmed
+    }
+}
+
+public interface IShoppingCartRepository
+{
+    ValueTask<ShoppingCartModel?> Find(Guid id, CancellationToken ct);
+
+    Task Store(Guid id, ShoppingCartModel? current, ShoppingCartEvent[] events, CancellationToken ct);
+
+    async Task GetAndStore(Guid id, Func<ShoppingCartModel?, ShoppingCartEvent[]> handle, CancellationToken ct)
+    {
+        var state = await Find(id, ct);
+
+        var events = handle(state);
+
+        await Store(id, state, events, ct);
+    }
+}
+
+public class ShoppingClassCommandHandler(
+    ShoppingClassDecider decider,
+    IShoppingCartRepository repository
+)
+{
+    public Task Handle(Guid id, ShoppingCartCommand command, CancellationToken ct) =>
+        repository.GetAndStore(
+            id,
+            model => decider.Decide(command, Map(model)),
+            ct
+        );
+
+    private ShoppingCart Map(ShoppingCartModel? model)
+    {
+        if (model == null)
```

Comment:
could make it an expression, i.e. `=> model = null ? new Empty () : model.Status switch { ...` ?

Reply:
I tried and wan’t able to do such.

Reply:
if the throw helper https://github.com/oskardudycz/latest-csharp-for-domain/pull/1#discussion_r1392618090 can take the value...
looking froward to total matching / sealed hierarchies!